presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestOrcFileRewriter.java

```diff
@@ -413,9 +428,225 @@ public void testUncompressedSize(boolean useOptimizedOrcWriter)
+                TupleDomain.all(),
+                READER_ATTRIBUTES);
+
+        Page page = null;
+        while (page == null) {
+            page = source.getNextPage();
+        }
+        assertEquals(page.getPositionCount(), 1);
+
+        // Column 13L
+        Block column0 = page.getBlock(0);
+        assertTrue(column0.isNull(0));
+
+        // Column 7L
+        Block column1 = page.getBlock(1);
+        assertEquals(createVarcharType(20).getSlice(column1, 0), utf8Slice("2"));
+
+        // Column 8L
+        Block column2 = page.getBlock(2);
+        assertTrue(column2.isNull(0));
+
+        dummyHandle.close();
+        deleteRecursively(dataDir.toPath(), ALLOW_INSECURE);
+    }
+
+    /**
+     * The following test drop and add the same columns; the legacy ORC rewriter will fail due to unchanged schema.
+     * However, if we enforce the newly added column to always have the largest ID, this won't happen.
+     */
+    @Test(dataProvider = "useOptimizedOrcWriter")
+    public void testRewriterDropThenAddSameColumns(boolean useOptimizedOrcWriter)
+            throws Exception
+    {
+        TypeRegistry typeRegistry = new TypeRegistry();
+        DBI dbi = new DBI("jdbc:h2:mem:test" + System.nanoTime());
+        dbi.registerMapper(new TableColumn.Mapper(typeRegistry));
+        Handle dummyHandle = dbi.open();
+        File dataDir = Files.createTempDir();
+
+        StorageManager storageManager = createOrcStorageManager(dbi, dataDir, useOptimizedOrcWriter);
+
+        List<Long> columnIds = ImmutableList.of(3L, 7L);
+        List<Type> columnTypes = ImmutableList.of(BIGINT, createVarcharType(20));
+
+        File file = new File(temporary, randomUUID().toString());
+        try (FileWriter writer = createFileWriter(columnIds, columnTypes, file, false, useOptimizedOrcWriter)) {
+            List<Page> pages = rowPagesBuilder(columnTypes)
+                    .row(2L, "2")
+                    .build();
+            writer.appendPages(pages);
+        }
+
+        // Add a column
+        File newFile1 = new File(temporary, randomUUID().toString());
+        OrcFileInfo info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(3L, 7L, 10L), ImmutableList.of(BIGINT, createVarcharType(20), DOUBLE)),
+                file,
+                newFile1,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Drop a column
+        File newFile2 = new File(temporary, randomUUID().toString());
+        info = createFileRewriter(useOptimizedOrcWriter).rewrite(
+                getColumnTypes(ImmutableList.of(7L, 10L), ImmutableList.of(createVarcharType(20), DOUBLE)),
+                newFile1,
+                newFile2,
+                new BitSet(5));
+        assertEquals(info.getRowCount(), 1);
+
+        // Add a column with the same ID but different type
```

Comment:
See my comments above. I think preventing reusing the same column ID is a separate issue. We should never reuse column id for a table.

Reply:
As discussed offline, let's keep it to raise the awareness of the buggy legacy rewriter.